lls/main.c

```diff
@@ -35,15 +48,53 @@ get_lls_conf(void)
 static int
 cleanup_lls(void)
 {
-	/* TODO Destroy ARP and ND cache. */
+	lls_cache_destroy(&lls_conf.arp_cache);
+	/* TODO Destroy ND cache. */
 	rte_timer_stop(&lls_conf.timer);
 	return 0;
 }
 
+int
+arp_resolve(struct in_addr *ipv4_addr, struct ether_addr *eth_addr)
+{
+	return lls_cache_get(&lls_conf.arp_cache, ipv4_addr, eth_addr);
+}
+
+int
+arp_hold(struct in_addr *ipv4_addr, uint8_t port_id)
+{
+	return lls_cache_refcnt(&lls_conf.arp_cache, ipv4_addr, port_id, true);
+}
+
+int
+arp_put(struct in_addr *ipv4_addr, uint8_t port_id)
+{
+	return lls_cache_refcnt(&lls_conf.arp_cache, ipv4_addr, port_id, false);
+}
+
+static inline size_t
+get_vlan_offset(struct ether_hdr *eth_hdr, uint16_t *proto)
+{
+	size_t vlan_offset = 0;
+	if (rte_cpu_to_be_16(ETHER_TYPE_VLAN) == *proto) {
+		struct vlan_hdr *vlan_hdr = (struct vlan_hdr *)(eth_hdr + 1);
+
+		vlan_offset = sizeof(struct vlan_hdr);
+		*proto = vlan_hdr->eth_proto;
+
+		if (rte_cpu_to_be_16(ETHER_TYPE_VLAN) == *proto) {
```

Comment:
I've taken the VLAN tag handling out of the patch entirely, since I added it as a TODO in an earlier patch. Otherwise, LLS would be the only block handling the VLAN.

So until that TODO is done, the assumption in the code is that there is no VLAN tags at all.